Included sending string and also initializing baudrate

Kevin-L-Dai opened the pull request:
I added a "send string" function and also added baudRate into the initUART.

Kevin-L-Dai committed:
Included sending string and also initializing baudrate

I added a "send string" function and also added baudRate into the initUART.

cecilsmith commented on 2_Code/src/uart.c:
```diff
@@ -1,20 +1,95 @@
+    
+    U1MODEbits.UARTEN = 0;
+    U1MODEbits.BRGH = 0;
+    U1BRG = baudRate; // 16 MHz / (16 * 9600) - 1 = 103.16666667
```
Perhaps it would be helpful to keep the calculation. Otherwise, it would be confusing for the user if they have to input 103 as opposed to 9600.

cecilsmith commented on 2_Code/src/uart.c:
```diff
@@ -1,20 +1,95 @@
 }
 
-void sendPackage(char data[])
+void init_UART(baudRate)
```
We should specify what `baudRate` is (likely `unsigned int`).

cecilsmith commented on 2_Code/src/uart.h:
```diff
@@ -1,22 +1,26 @@
+void send_GPS_Str_command(unsigned char* s);
+void send_GPS_Char_command(unsigned char command);
+unsigned char get_GPS_data(void);
+void init_UART(baudRate);
```
Same as earlier comment, should specify what `baudRate` is (likely `unsigned int`).

cecilsmith commented on 2_Code/src/uart.c:
```diff
@@ -1,20 +1,95 @@
+    // Configures UART module, pins, and interrupt flags
+
+    CLKDIVbits.RCDIV = 0;
+    AD1PCFG = 0x9fff;  // For digital I/O.
```
Although it is more work, we should clarify which pins we are using for UART as opposed to overwriting all configurations from other `setup()`/`init()` functions.

cecilsmith replied:
@Kevin-L-Dai, is there a way we can set up `AD1PCFG` for our specified pins for UART?

Kevin-L-Dai replied:
we could set up a AD1PCFG for our specific pins for UART however it isn't entirely necessary since AD1PCFG is specifically for the pins labeled as AN0 - AN12. Since the pins I am using 
(_TRISB6 = 0;  // Sets pin 15 (RB6) to U1TX output pin
    _TRISB10 = 1; // Sets pin 21 (RB10) to U1RX input pin
)
are not on any AN pins of the PIC24, it most likely isn't necessary to do any AD1PCFG stuff. At least for UART. At least thats what I understood from my TA. He did suggest though that if we aren't doing any analog readings, we could just set all the pins to digital just in case.

<img width="684" alt="Screenshot 2024-04-08 at 1 37 04 PM" src="https://github.com/cecilsmith/GPS-MINPOINT/assets/163371180/51ae0407-23ac-4080-beac-576fcef8d47b">

cecilsmith commented on 2_Code/src/uart.c:
```diff
@@ -1,20 +1,95 @@
-    // Peripheral Pin Select
+    while (front == back) ;
+    NMEA = buffer[back++];
+    back &= 63;
```
Cool algorithm, but for our understanding, could you clarify why we use `back &= 63;`?
Thanks!

cecilsmith commented on 2_Code/src/uart.c:
```diff
@@ -1,20 +1,95 @@
-    U1MODEbits.BRGH = 0; // Not high speed
-    U1BRG = (unsigned int)FCY / (16 * baudRate) - 1; // Calculate baud (should be 103)
-    U1MODEbits.ON = 1;   // Enable UART
+unsigned char get_GPS_data(void)
```
If we get more than one `char` back from the GPS data, lets say a string, how will this work? Could we implement a String getter?

cecilsmith reviewed (changes requested):
Love the work on this. Looks pretty good so far, just a few things that could use some adjustment :)

cecilsmith commented:
Also, for clutter purposes, it may also be a good idea to remove the `.DS_Store` files.

Kevin-L-Dai committed:
Implementations of the suggestions

I have changed the baudrate calculation so that we can enter 9600 and it automatically calculates it to 103.

I have also added the unsigned int before baudrate. I was wondering why it was giving an error lol.

as for the setup/init pin initialization, i removed setting all pins to digital and just kept setting the pins we need (_TRISB6 = 0; _TRISB10 = 1;)

as for back &= 63, that was originally copied from the discussion to implement a circular buffer. however, it didn't work when i tried implementing it in my lab so i have upddated it to a version that works and is self explanatory.

Finally, I have added a string getter. It requires a pointer string  and also the expected size of the string.

Kevin-L-Dai committed:
Delete .DS_Store

Kevin-L-Dai committed:
Delete 2_Code/src/.DS_Store

Kevin-L-Dai committed:
Delete 2_Code/.DS_Store

Kevin-L-Dai reviewed (commented):
I have addressed the issues you have brought up. I don't know if my description for the commit is visible somewhere but I have described what I have done there.

cecilsmith commented:
Okay. Two small(er) things.

1. We should be careful about returning a `char` if/when there isn't logic to return said `char` within the function.
2. We should make sure that the UART library works with a tool like PuTTY (alternatively, we could test with an Arduino).

Kevin-L-Dai committed:
Small fix of the void

I have changed the unsigned char to void on that one function we looked at together.

Kevin-L-Dai commented:
I do not know how to use PuTTY. However, I have asked the undergrad student assistant here if the UART Library looked like it was good to go and functional. He said it looked good so that may be a bit of reassurance. It would still probably be a good idea to test it though.

Kevin-L-Dai commented:
I will try to get a start on the next part of the project (gps.c, gps.h) assuming that the library works.

cecilsmith committed:
Merge branch 'Second_UART_Library_withStr' of https://github.com/cecilsmith/GPS-MINPOINT into Second_UART_Library_withStr

cecilsmith reviewed (approved):
Looks great. Thanks for all the work!
